### DISCUSSION TITLE
openEMS/Blender simulations and renders showcase

### DISCUSSION THREAD

**Discussion by samerps** [thliebig/openEMS-Project#48] [Show and tell]
Hi all, I would like to start a thread here to showcase my work with openEMS and Blender. I have started using openEMS last month and it has been a fantastic tool to simulate and understand more about electromagnetics. I also have been using Blender over the last two years to create renders and animations of electronics. And so naturally the next thing is to use the results from openEMS and render them in a Blender scene. My goal is to make it simpler and fun for people to understand all things related to electromagnetics and circuits, and hopefully inspire people to study these topics. 

I am planning to share my workflow, models and tools in the near future. So far my workflow is as follows:
- create geometry and meshes in Blender 
- export geometry as STL files
- setup openEMS simulation and import STL files 
- save simulation data as vtk files 
- open vtk in paraview and perform any necessary filter operations such as clipping and warp
- export paraview scene as x3d files via a Python script 
- import x3d files into Blender via a Python script
- use Blender's Geometry Nodes system to visualise and animate

My plan is to skip paraview, and import vtk files directly into Blender. I have looked at [BVtkNodes](https://github.com/tkeskita/BVtkNodes) for Blender but didn't have much success in get the data imported correctly. 

-----

My first render/animation is a simple one, showing what happens to an electromagnetic wave when it is launched into a coax cable with an open termination. 
You can see the animation in 4k on YouTube 
[https://www.youtube.com/watch?v=okZ8GFbtjpo](https://www.youtube.com/watch?v=okZ8GFbtjpo)

![render1](https://user-images.githubusercontent.com/68449711/227720181-5a2d6e80-aea5-484a-b389-0ec0ea2351d3.jpg)
![render2](https://user-images.githubusercontent.com/68449711/227720183-6b86ee01-38ff-4918-bd14-82831981b585.jpg)

I will also be publishing these renders on my social media accounts 
https://twitter.com/samerps
https://www.linkedin.com/in/samerps/
https://www.artstation.com/high_voltage

**Comment by thliebig** [comment]
Thanks for sharing. This is indeed something I have done in the past too and it is always a lot of fun to see the final result...

**Comment by samerps** [comment]
### 2.4Ghz patch antenna with a coax cable feed
My next simulation scene is for a 2.4Ghz patch antenna with a coax cable feed. I've shown the electric (blue) and magnetic fields (orange) just above the surface of the antenna, and you can see how the fields are orthogonal to each other and intense at different edges of the antenna. Really enjoyed working on this example and these animations helped me understand antennas a bit more. 

You can watch the video at HD resolution on YouTube 
https://www.youtube.com/watch?v=vABw3vBoVzk

https://user-images.githubusercontent.com/68449711/229171496-3df391db-dc63-49c2-91d6-8b878b2c45db.mp4

https://user-images.githubusercontent.com/68449711/229173803-7cc40939-e4f8-434e-9482-1249a2bf2ab9.mp4

![render3](https://user-images.githubusercontent.com/68449711/229171847-f539a1f9-c483-42d5-af92-646d3c435d86.jpg)
![render4](https://user-images.githubusercontent.com/68449711/229173838-f05fcb4a-7f20-418d-90a0-b57122edffbf.jpg)

**Reply by NathanGregKidd** [reply]
Hi! I've been super inspired by your work, and I've been trying it out myself. I'm wondering how you created that big antenna spread bubble in the last moments of the 2.4Ghz.Patch.Antenna.Simulation.openEMS.blender3d.mp4 video. 

Also, have you played with the NF2FF stuff in openEMS?

**Reply by samerps** [reply]
Hi Nathan, If I remember correctly I exported it from Paraview as x3d file. I have not experimented much with NF2FF

**Comment by sebgus** [comment]
Hi @samerps !

Excellent work. Looking forward to seeing the detailed workflow!

**Comment by samerps** [comment]
My next set of simulations and renders are for a 2.4GHz meandered inv F PCB antenna, based on the design from Texa's Instruments App note AN043 
In this animation you can see how the surface current is distributed in the PCB antenna

https://user-images.githubusercontent.com/68449711/234607632-860e0c78-8b51-473f-89e2-5b46910e6e4b.mp4

**Reply by samerps** [reply]
this animations shows the E-field component and how it radiates away from the antenna 

https://user-images.githubusercontent.com/68449711/235691649-3370b302-1044-4033-8f60-5a25db7926f5.mp4

**Comment by PinkMushroom** [comment]
> setup openEMS simulation and import STL files

How are you doing that ?!
I cannot find anything in the documentation that provides for STL import.

**Reply by samerps** [reply]
use the 'ImportSTL' function,
example:
`CSX = AddMetal( CSX, 'cad_model' ); % create a perfect electric conductor (PEC) `
`CSX = ImportSTL(CSX, 'cad_model',10, 'sphere.stl','Transform',{'Scale',1/unit});`

**Reply by warsea94** [reply]
I was able to import my STL using these commands. But I am confused as to how to give the Boundary conditions for this case of mine.
![cyl](https://github.com/thliebig/openEMS-Project/assets/144090498/5f47c342-7808-4025-b842-1dc971fe74ed)

I have a bent pipe (yellow ) which is inside a cylindrical tank (denoted by red). I want to have an excitation source inside the pipe and give PML BC on the surface of the pipe as shown in the image. 
![problem](https://github.com/thliebig/openEMS-Project/assets/144090498/47a84de5-9813-457f-ae04-605a209d966b)

I am not able to find anything in the documentation.

**Reply by LubomirJagos** [reply]
you can't add boundary condition to some shape, it's always around whole model, it's defined as this:
```matlab
BC = {"PEC","PEC","PEC","PEC","PEC","PEC"};
FDTD = SetBoundaryCond( FDTD, BC );
```

I gues all what you can do to get closes to PML is to define some material with high Kappa, something like this, but not sure if it will behave like matched layer:
```matlab
CSX = AddMaterial(CSX, 'high_loss_material');
CSX = SetMaterialProperty(CSX, 'high_loss_material', 'Epsilon', 1, 'Mue', 1, 'Kappa', 1e6);
```

**Comment by samerps** [comment]
PCB layout simulation. I was able to simulate the layout of a two layer FR4 PCB that has a meandered inv F antenna designed for 2.4GHz, I was interested to see how the routing and layout will impact S11. The result I got showed that my specific layout caused S11 to dip at 2.55GHz instead of 2.4GHz 

https://github.com/thliebig/openEMS-Project/assets/68449711/f0801037-c342-4472-87aa-ce41c78d121c

![s11plot](https://github.com/thliebig/openEMS-Project/assets/68449711/fb590e92-cd96-4ede-8393-a30007ea347a)

**Comment by XDAChen** [comment]
Hi Samer, 
 Very cool work! For my project, I am planning to do similar works (Blender + EM simulation).
I wonder how you take care of the port excitation when simulating a custom geometry as an STL file? Anything specific you suggest to pay attention to in order to maintain simulation accuracy? I want to simulate a conformally bent antenna made in Blender but not clear how should I accurately assign port excitations.

**Reply by samerps** [reply]
thanks! I have been using lumped ports, the only thing I do is making the port box overlap slightly with the STL mesh

**Reply by XDAChen** [reply]
Hi Samer,

I am getting my hands wet in openEMS + blender as well. I want to ask what was your meshing plan to maintain high simulation precision. In the examples you provided, the coax cable is canonically bent, and the entire board including the wire and the patch are simulated so I am curious how complex did you design your mesh grid?

Does openEMS only support cartesian mesh and cylindrical mesh? do you know if openEMS allows users to define more customized unstructured grids (triangle or tetrahedral) that can accompany an accurate simulation of a flexible antenna?

thanks!

**Reply by 0xCoto** [reply]
> do you know if openEMS allows users to define more customized unstructured grids (triangle or tetrahedral)

These types of meshes are not supported in openEMS/FDTD. Solvers like the finite element method (FEM) and method of moments (MoM) —whose meshes are tetrahedral— are sometimes a bit more suitable for complex geometries with curves, turns and such. Each solver has its advantages and disadvantages depending on the application.

**Reply by XDAChen** [reply]
> > do you know if openEMS allows users to define more customized unstructured grids (triangle or tetrahedral)
> 
> These types of meshes are not supported in openEMS/FDTD. Solvers like the finite element method (FEM) and method of moments (MoM) —whose meshes are tetrahedral— are sometimes a bit more suitable for complex geometries with curves, turns, and such. Each solver has its advantages and disadvantages depending on the application.

Thanks for the info. In fact, I switched from HFSS to openEMS because HFSS is not friendly towards auto-simulation with canonical surfaces. I have computing power so if I get my mesh small enough my guess is that it would render decent accuracy even if the surface is complex. Another tool I can try is MATLAB antenna toolbox that uses MoM but openEMS seems to offer more flexibility.

**Reply by samerps** [reply]
Hi @XDAChen, the mesh for the 'bent' coax demo is very simple, have a look at the screenshots below. 
![image](https://github.com/thliebig/openEMS-Project/assets/68449711/0c8a4a8c-e44b-43d2-8ad4-b48e8c7cb19d)


[BentCoax.zip](https://github.com/thliebig/openEMS-Project/files/11584407/BentCoax.zip)
c2a7ec)

**Reply by XDAChen** [reply]
Thank you @samerps . Your mesh set up was insightful, I am still working on my simulation pipeline and would love to share it when it is competed.

**Comment by samerps** [comment]
### 60GHz mmWave Simulations!

I will be using one Texas Instruments mmWave evaluation boards, specifically TIDEP-01001 which features the AWR6843 60GHz single-chip mmWave sensor. The design consists of a 60GHz transmitting and receiving patch antenna arrays that are series-fed from transmission lines to and from the AWR6843 sensor. 
This first simulation result shows the electric field component of the wave propagates from the sensor to the patch antenna for a single channel, I’ve drawn the simulated field outside for better visualization. The S11 results shows dips at 60.4GHz and 63GHz which seems to agree with what was described in the design guide. Next, I’ll be looking at driving all the channels to see radiation pattern, and possibly see how objects interact with mmwaves

https://github.com/thliebig/openEMS-Project/assets/68449711/b70819cb-b79b-4cd3-a209-8622cf6b9f1b

![Uploading s11_singlechannel.jpg…]()

**Reply by samerps** [reply]
very excited to share this new animation showing all four transmission lines and the radiated E-Field. I am very impressed with the capabilities of openEMS of simulating mmWave PCB designs and with Blender’s capability of handling large amounts of data.  Note the simulated patch antenna array is meant to be the receiver in the actual design.  

https://github.com/thliebig/openEMS-Project/assets/68449711/c95309e4-cf1a-4b83-9eea-fd7f23fedb3f

**Reply by samerps** [reply]
This is how a human finger interacts with 60GHz mm waves. I assumed the em waves launched from the patch antenna array become a planar wave as it reaches the finger. The hand and finger was modelled in Blender and exported as STL into openEMS

https://github.com/thliebig/openEMS-Project/assets/68449711/9743fa26-d66d-4a6e-aecd-20113b6d7037

**Comment by samerps** [comment]
The apollo lunar rover had a high gain antenna with a foldable parabolic reflector for voice and video transmission to earth. This simulation shows how the reflector dish unfolds and how the received electromagnetic waves become focused at the antenna feed

https://github.com/thliebig/openEMS-Project/assets/68449711/0fa572ce-c503-4632-b680-43d6aec9f4d0

**Reply by samerps** [reply]
https://github.com/thliebig/openEMS-Project/assets/68449711/5e9d7b28-7020-47be-bd8f-54a92e3774f1

**Reply by toammann** [reply]
Hey,
this is nice. One can clearly see how the reflector collects and reflects the wave and focuses it on the feed :)

**Reply by KJ7LNW** [reply]
These are amazing.  I can't wait to see some automated tooling for this and play with it.  

The dish antenna signals look polarized.  Would it be a CP signal in real life?

Can you model a CP antenna excitation?

**Comment by samerps** [comment]
Attempting to simulate a 70GHz Rotman Lens with openEMS! After seeing a few inspiring tweets from 
[@QVHenkel](https://twitter.com/QVHenkel) and  [@ptrschmdtnlsn](https://twitter.com/ptrschmdtnlsn) I thought I'd have a go at it. 

https://www.youtube.com/watch?v=L-mRn4BJroo
![thumbnail](https://github.com/thliebig/openEMS-Project/assets/68449711/5254dd29-aad2-429f-a51d-03da3040e737)


This animation shows the launch of the electric field component from one port and how it propogates throughout the lens. It does seem that most of waves are absorbed at each inactive port, though there are some reflected waves which I believe may be due to the mesh of the model.

I will do another sim with the phased array elements to look how beamforming is achieved.

**Comment by ALEEF02** [comment]
@samerps Like a few before have asked, I'm looking for your documented workflow. Please let me know where I can find such. THe more detail the better! :D

**Reply by samerps** [reply]
Hi, I have not documented my work flow, but I did publish a 45min YouTube video where you can see my workflow, see below
https://github.com/thliebig/openEMS-Project/discussions/214

**Reply by ALEEF02** [reply]
@samerps Would you be able to publish your files for the Rotman lens? I would like to investigate your mesh, as I'm having troubles with my own.

**Reply by samerps** [reply]
sure, I've uploaded them in the original post #214
